### PR TITLE
(FACT-3140) Include partition type uuid for GPT based systems

### PR DESCRIPTION
**Pull request by jcpunk**
I've taken a start at getting GPT GUID types into facter.

I fear I'll need a bit of help getting it over the line, but it should be a workable starting point....

### CODE REVIEW

File: lib/facter/resolvers/partitions.rb [hunk 1]
```diff
@@ -121,7 +121,7 @@ def execute_and_extract_blkid_info
         def populate_from_lsblk(partition_name, blkid_and_lsblk)
           return {} unless available?('lsblk', blkid_and_lsblk)
 
-          blkid_and_lsblk[:lsblk] ||= Facter::Core::Execution.execute('lsblk -fp', logger: log)
+          blkid_and_lsblk[:lsblk] ||= Facter::Core::Execution.execute('lsblk -p -P -o NAME,FSTYPE,LABEL,UUID,PARTTYPE', logger: log)
```
[the remark this answers is not in the document]
Comment (jcpunk) [reply lib/facter/resolvers/partitions.rb:124]: I fear `PARTTYPE` is what I was hopeful to capture with this patch set....

[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/facter/resolvers/partitions.rb:124]: It looks like `PARTTYPE` was added in https://github.com/util-linux/util-linux/commit/4686ba17a8239d9a840a901abf780c4ea02ea1f9 and first released in 2.25. So we'd need to make this conditional based on the version.

[the remark this answers is not in the document]
Comment (jcpunk) [reply lib/facter/resolvers/partitions.rb:124]: I'm not sure how to do that....

[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/facter/resolvers/partitions.rb:124]: You'll want to do something like what this resolver does https://github.com/puppetlabs/facter/blob/39b48ad06e8dd87b8ef6d291c1c5588d7499ab36/lib/facter/resolvers/augeas.rb#L32

to detect the version and only include the new fact on more recent versions.

[the remark this answers is not in the document]
Comment (jcpunk) [reply lib/facter/resolvers/partitions.rb]: I took a stab at implementation.

**Comment by jcpunk** [comment]
Rebased off HEAD

**Comment by joshcooper** [comment]
@jcpunk you'll need to stub out the call to `lsblk --version 2>&1` similar tohttps://github.com/puppetlabs/facter/blob/465c119f1ed7aeb8f1b2750e07399f0f251d87ab/spec/facter/resolvers/partitions_spec.rb#L44-L45

Also please add a test to test the fallback behavior when the version is older and verify facter omits the partition type.

**Comment by jcpunk** [comment]
Long term it might make more sense to just run the version with the extra options and have it catch the error if it isn't supported.

It may be a good while before I get back to this.

**Comment by mhashizume** [comment]
I've started digging into this and I think the way the partition resolver currently works, in most cases the resolver won't even use `lsblk`. The partitions resolver will always try to get the output from `blkid` first and will only use `lsblk` as a fallback:

https://github.com/puppetlabs/facter/blob/e776b42608ce9f5be49805520ca81f0e125a674a/lib/facter/resolvers/partitions.rb#L76-L78

However, it seems like `lsblk` can return all of the information that `blkid` can and more. It might make sense to prefer `lsblk` over `blkid` in the resolver and add `PARTUUID` to `lsblk`'s output to reach parity with what we already return with `blkid`.

Also a nit, but I think "GUID" is preferred over "UUID" in this context. The UEFI spec refers to what we're discussing here as "PartitionTypeGUID" (page 125): https://uefi.org/sites/default/files/resources/UEFI_Spec_2_9_2021_03_18.pdf

I can work more on this next week and put up my own PR if you can't get back to this for a while.

**Comment by jcpunk** [comment]
If you're able to put one together that would be great!

**Comment by mhashizume** [comment]
Closing this in favor of https://github.com/puppetlabs/facter/pull/2745

**Pull request closed without merging**